lib/datasource.js

```diff
@@ -799,16 +809,24 @@ DataSource.prototype.automigrate = function (models, cb) {
     });
 
     if (invalidModels.length) {
-      return process.nextTick(function () {
+      var error = process.nextTick(function () {
         if (cb) {
           cb(new Error('Cannot migrate models not attached to this datasource: ' +
           invalidModels.join(' ')));
         }
       });
+
+      if (cb.promise) {
+        return cb.promise;
+      }
+      return error;
     }
   }
 
   this.connector.automigrate(models, cb);
+  if (cb.promise) {
+    return cb.promise;
+  }
```

Comment:
It's not necessary to check if `cb.promise` is truthy, you can always `return cb.promise`. The result will be the same(*):
 - When `cb.promise` is not defined, the return value will be `undefined`.
 - When a function does not return anything (does not call `return`), then the return value is `undefined `too.

(*) Exception: if `cb.promise` is defined and falsy (e.g. `0`, `null` or `''`), your version returns `undefined` while my version returns the falsy value. I think this difference does not really matter.